pyro/nn/module.py

```diff
@@ -0,0 +1,196 @@
+"""
+Pyro includes a special class :class:`~pyro.nn.module.PyroModule` whose
+attributes can be modified by Pyro effects. To create one of these attributes,
+use either the :class:`PyroParam` struct or the :class:`PyroSample` struct::
+
+    my_module = PyroModule()
+    my_module.x = PyroParameter(torch.tensor(1.), constraint=constraints.positive)
+    my_module.y = PyroSample(dist.Normal(0, 1))
+
+"""
+from collections import OrderedDict, namedtuple
+
+import torch
+from torch.distributions import constraints, transform_to
+
+import pyro
+import pyro.distributions as dist
+
+PyroParam = namedtuple("PyroParam", ("init_value", "constraint", "event_dim"))
+PyroParam.__new__.__defaults__ = (constraints.real, None)
+PyroParam.__doc__ = """
+Structure to declare a Pyro-managed learnable parameter of a :class:`PyroModule`.
+"""
+
+PyroSample = namedtuple("PyroSample", ("prior",))
+PyroSample.__doc__ = """
+Structure to declare a Pyro-managed random parameter of a :class:`PyroModule`.
+"""
+
+
+def _make_name(prefix, name):
+    return "{}.{}".format(prefix, name) if prefix else name
+
+
+class PyroModule(torch.nn.Module):
+    """
+    Subclass of :class:`torch.nn.Module` that supports setting of
+    :class:`PyroParam` and :class:`PyroSample` .
+
+    To create a Pyro-managed parameter attribute, set that attribute using the
+    :class:`PyroParam` helper. Reading that attribute will then trigger a
+    :func:`pyro.param` statement. For example::
+
+        # Create Pyro-managed parameter attributes.
+        my_module = PyroModule()
+        my_module.loc = PyroParam(torch.tensor(0.))
+        my_module.scale = PyroParam(torch.tensor(1.),
+                                    constraint=constraints.positive)
+        # Read the attributes.
+        loc = my_module.loc  # Triggers a pyro.param statement.
+        scale = my_module.scale  # Triggers another pyro.param statement.
+
+    Note that, unlike normal :class:`torch.nn.Module` s, :class:`PyroModule` s
+    should note be registered with :func:`pyro.module` statements.
+    :class:`PyroModule` s can contain normal :class:`torch.nn.Module` s, but not
+    vice versa. Accessing a normal :class:`torch.nn.Module` attribute of
+    a :class:`PyroModule` triggers a :func:`pyro.module` statement.
+
+    To create a Pyro-managed random attribute, set that attribute using the
+    :class:`PyroSample` helper, specifying a prior distribution. Reading that
+    attribute will then trigger a :func:`pyro.sample` statement. For example::
+
+        # Create Pyro-managed random attributes.
+        my_module.x = PyroSample(dist.Normal(0, 1))
```

Comment:
I like the interface and the fact that `PyroParam` is poutine aware unlike `ConstrainedParam`. Could you talk about the use case for `PyroSample` as attributes? 

Unlike `params` that are useful as attributes that can be optimized and updated, `sample` statements in most models have a more limited scope and are mostly useful when aggregated. I suppose the biggest difference here is that we don't need to name them. Do you also foresee using these to collect posterior samples from a trained module?

Reply:
Okay, I see Martin's PR as being the motivation for this. 🙂

Reply:
As other motivation: consider using Pyro to train a 3rd-party `nn.Module`, say treating it as a Bayesian nn. We would point-estimate some `nn.Parameter`s by translating them to `PyroParam`s, and be fully Bayesian about other parameters, by translating them to `PyroSample` statements with given prior.

It remains to be seen how we can do this post-hoc, since we would need to insert `PyroModule` in the 3rd-party inheritance hierarchy. An intrusive partial approach is to replace `nn.Module` with `pyro.nn.PyroModule` in their code. But I'd prefer to find a way to dynamically change the type of objects in an existing data structure, similar to our fancy `Distribution.expand()` logic. E.g., we could temporarily override `nn.Module.__new__()` to instead return a `pyro.nn.PyroModule`.